common/junit-platform-native/src/main/java/org/graalvm/junit/platform/JUnitPlatformFeature.java

```diff
@@ -156,20 +157,31 @@ private TestPlan registerTestPlan(Launcher launcher, List<Path> classpath) {
                 .map(Optional::get)
                 .filter(ClassSource.class::isInstance)
                 .map(ClassSource.class::cast)
-                .forEach((ClassSource classSource) -> {
-                    Class<?> clazz = classSource.getJavaClass();
-                    if (debug) {
-                        System.out.println("[Debug] Found test class: " + clazz);
-                    }
-                    RuntimeReflection.registerForReflectiveInstantiation(clazz);
-                    for (Field field : clazz.getDeclaredFields()) {
-                        RuntimeReflection.register(field);
-                    }
-                    for (Method method : clazz.getDeclaredMethods()) {
-                        RuntimeReflection.register(method);
-                    }
-                });
+                .forEach(this::registerTestClassForReflection);
 
         return testPlan;
     }
+
+    private void registerTestClassForReflection(ClassSource classSource) {
+        Class<?> clazz = classSource.getJavaClass();
+        if (debug) {
+            System.out.println("[Debug] Found test class: " + clazz);
+        }
+
+        Class<?> declaringClass = clazz.getDeclaringClass();
+        if (declaringClass == null || Modifier.isStatic(clazz.getModifiers())) {
+            /* A top level or a static nested class must have a no-arg constructor. */
+            RuntimeReflection.registerForReflectiveInstantiation(clazz);
+        } else {
+            /* A non-static nested class cannot have a no-arg constructor as it needs a reference to an enclosing parent's object. */
+            RuntimeReflection.register(clazz.getDeclaredConstructors());
```

Comment:
We don't need to explicitly register the declaring class for reflection here as it will be one of the processed test classes.